log/logtest/assert.go

```diff
@@ -0,0 +1,63 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package logtest // import "go.opentelemetry.io/otel/log/logtest"
+
+import (
+	"context"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+
+	"go.opentelemetry.io/otel/log"
+)
+
+// TestingT reports failure messages.
+// [testing.T] implements this interface.
+type TestingT interface {
+	Errorf(format string, args ...any)
+}
+
+// AssertEqual asserts that the two concrete data-types from the logtest package are equal.
+func AssertEqual[T Recording | Record](t TestingT, want, got T, opts ...AssertOption) bool {
```

Comment:
Why not 
```suggestion
func AssertEqual[T Recording | Record](t *testing.T, want, got T, opts ...AssertOption) bool {
```

Won't this break if we add methods to `TestingT`?

Reply:
So that it can be reused in fuzz tests and benchmarks.

We can use type assertions instead of adding a new method to this type definition. This is also follows https://pkg.go.dev/github.com/stretchr/testify/assert#TestingT and what testify did e.g. to support t.Helper

Reply:
Moreover, it makes it possible to test the test failures 😄

Reply:
> So that it can be reused in fuzz tests and benchmarks.

But we aren't doing either of these here, right?

Reply:
> Moreover, it makes it possible to test the assertion failures 😄

What assertion failures?

Reply:
I mean tests in `log/logtest/assert_test.go` where `want` is `false`.

Reply:
I see that it is also similar to:
- https://github.com/open-telemetry/opentelemetry-go/blob/7639c93baffc820f9f19ab1f9d1c085b0b45e071/sdk/metric/metricdata/metricdatatest/assertion.go#L47-L59

Reply:
I think the `metricdatatest` package speaks to the issues I see where we documented that we cannot add new methods to that type. If the upstream `testing.T` type has new methods added to it we would be able to use them directly without type assertions if we accept the concrete type itself.

Accepting a `TestingT` type is going to be helpful if we expect to accept multiple implementations of that interface. But I don't see that happening with how this is currently designed and scoped.

Reply:
I'm interested to hear what other @open-telemetry/go-approvers think.

Reply:
also @khushijain21, @scorpionknifes

Reply:
> So that it can be reused in fuzz tests and benchmarks.

I see this  a good reason to keep `TestingT` interface - we may have a usecase for AssertEqual(..) in benchmarks/fuzz tests

Reply:
Benchmarks shouldn't have to perform assertions, as that skews their results.
Being able to reuse those assertions for fuzz testing would be nice, though I'm not sure that'd be so useful (the nature of fuzz testing makes exact assertions hard to write).

Reply:
@dmathieu, what is your preference then? :wink: Couple to `testing.T` or have the `TestingT` interface?

Reply:
I think the signature is more readable if we stick to `*testing.T`.

Reply:
Done. I also removed unused code.